lib/preflight/preflighttest.js

```diff
@@ -0,0 +1,483 @@
+              collectedStats.rtt.push(activeIceCandidatePair.totalRoundTripTime * 1000);
+            }
+            if (typeof activeIceCandidatePair.availableIncomingBitrate === 'number') {
+              collectedStats.incomingBitrate.push(activeIceCandidatePair.availableIncomingBitrate);
+            }
+
+            if (!collectedStats.selectedIceCandidatePairStats) {
+              collectedStats.selectedIceCandidatePairStats = {
+                localCandidate: activeIceCandidatePair.localCandidate,
+                remoteCandidate: activeIceCandidatePair.remoteCandidate
+              };
+            }
+          }
+
+          let packetsLost = 0;
+          let packetsReceived = 0;
+          if (remoteAudioTrackStats && remoteAudioTrackStats[0]) {
+            collectedStats.jitter.push(remoteAudioTrackStats[0].jitter);
+            packetsLost += remoteAudioTrackStats[0].packetsLost;
+            packetsReceived += remoteAudioTrackStats[0].packetsReceived;
+          }
+          if (remoteVideoTrackStats && remoteVideoTrackStats[0]) {
+            packetsLost += remoteVideoTrackStats[0].packetsLost;
+            packetsReceived += remoteAudioTrackStats[0].packetsReceived;
+          }
+          collectedStats.packetLoss.push(packetsReceived ? packetsLost * 100 / packetsReceived : 0);
+        }
+      });
+  }
+
+  function generatePreflightReport(collectedStats) {
+    testTiming.stop();
+    const selectedIceCandidatePairStats = collectedStats.selectedIceCandidatePairStats;
+    const isTurnRequired = selectedIceCandidatePairStats.localCandidate.candidateType === 'relay'
+    || selectedIceCandidatePairStats.remoteCandidate.candidateType === 'relay';
+
+    return {
+      roomSid: subscriberRoom.sid,
+      mediaRegion: subscriberRoom.mediaRegion,
+      signalingRegion: subscriberRoom.localParticipant.signalingRegion,
+      testTiming,
+      networkTiming: {
+        connect: connectTiming,
+        media: mediaTiming,
+      },
+      stats: {
+        jitter: makeStat(collectedStats.jitter),
+        rtt: makeStat(collectedStats.rtt),
+        outgoingBitrate: makeStat(collectedStats.outgoingBitrate),
+        incomingBitrate: makeStat(collectedStats.incomingBitrate),
+        packetLoss: makeStat(collectedStats.packetLoss),
+        networkQuality: makeStat(collectedStats.networkQuality),
+      },
+      selectedIceCandidatePairStats,
+      isTurnRequired,
+      iceCandidateStats: collectedStats.iceCandidateStats
+    };
+  }
+
+  function collectRTCStatsForDuration(duration, collectedStats = null) {
+    const startTime = Date.now();
+    const STAT_INTERVAL = 1000;
+    if (collectedStats === null) {
+      collectedStats = {
+        jitter: [],
+        rtt: [],
+        outgoingBitrate: [],
+        incomingBitrate: [],
+        packetLoss: [],
+        selectedIceCandidatePairStats: null,
+        iceCandidateStats: [],
+      };
+    }
+    return waitForSometime(STAT_INTERVAL).then(() => {
```

Comment:
This is a pause correct? Why do we need this? Will this affect time measurements?

Reply:
this ensures that we sample stats at 1000 ms. This is similar to doing something like:
```
setInterval(() => collectStats(), 1000)
```